source/common/http/headers.h

```diff
@@ -5,15 +5,48 @@
 #include "envoy/http/header_map.h"
 
 #include "common/singleton/const_singleton.h"
+#include "common/singleton/threadsafe_singleton.h"
 
 namespace Envoy {
 namespace Http {
 
+// This class allows early override of the x-envoy prefix from bootstrap config,
+// so that servers can configure their own x-custom-string prefix.
+//
+// Once the HeaderValues const singleton has been created, changing the prefix
+// is disallowed. Essentially this is write-once then read-only.
+class PrefixValue {
+public:
+  const char* prefix() {
+    absl::WriterMutexLock lock(&m_);
+    read_ = true;
+    return prefix_.c_str();
+  }
+
+  // The char* prefix is used directly, so must be available for the interval where prefix() may be
+  // called.
+  void setPrefix(const char* prefix) {
+    absl::WriterMutexLock lock(&m_);
+    RELEASE_ASSERT(!read_ || prefix_ == std::string(prefix),
```

Comment:
I'm sorry to be completely dense here, but why do we need the check to see if the prefix is being set to the same value? Can't we just check to see if it was ever read, and then just universally set on the next line?

Reply:
I guess this is just needed for the integration test, right? Maybe add a comment about that? It would be cool if there was a way to avoid this in product code just to allow the test, but if you can't think of anything easy I think a comment is fine. 

/wait

Reply:
Yeah, either we need the teardown in the integration test, or we need to avoid double setting the string here.
I'd prefer we take this out and have the extra code in TearDown, but you previously objected to that.

Reply:
OK got it. Sorry for the runaround I think this is all my confusion. I'm fine with this with the comment, but if you prefer the other way, feel free to swap and just add an equally sufficient comment trail?